docs/library/network.rst

```diff
@@ -29,6 +29,95 @@ For example::
+        provided. Most other methods require an active interface.
+
+    .. method:: connect([service_id, key=None, \*, ...])
+
+       Connect interface to the network. This method is optional, and
+       available only for interfaces which are not "always connected".
+       If no parameters are given, connect to the default (or only)
+       service. If a single parameter is given, it is primary identifier
+       of a service to connect to. It may be accompanied by a key
+       (password) required to access service. There can be further
+       arbitrary keyword-only parameters, depending on networking medium
+       type and/or particular device. Parameters can be used to: a)
+       specify alternative service identifer types; b) provide additional
+       connection parameters. For various medium types, there are different
+       sets of predefined/recommended paramaters, among them:
+
+       * WiFi: `bssid` keyword to connect by BSSID (MAC address) instead
+         of access point name
+
+    .. method:: disconnect()
+
+       Disconnect from network.
+
+    .. method:: isconnected()
+
+       Returns True if connected to network.
+
+    .. method:: scan(\*, ...)
+
+       Scan for the available network services/connections. Returns
+       list of tuples with discovered service parameters. For various
+       network media, there are different variants of predefined/
+       recommended tuple formats, among them:
+
+       * WiFi: (ssid, bssid, channel, RSSI, authmode, hidden). There
+       may be further fields, specific to a particular device.
+
+       The may accept additional keyword arguments to filter scan results
+       (e.g. scan for a particular service, on a particular channel,
+       for services of a particular set, etc.), and to affect scan
+       duration and other parameters. Where possible, parameter names
+       should match those in connect().
+
+    .. method:: status()
+
+       Return detailed status of the interface, values are dependent
+       on the network medium/technology.
+
+    .. method:: ifconfig([(ip, subnet, gateway, dns)])
+
+       Get/set IP-level network interface parameters: IP address, subnet mask,
+       gateway and DNS server. When called with no arguments, this method returns
+       a 4-tuple with the above information. To set the above values, pass a
+       4-tuple with the required information.  For example::
+
+        nic.ifconfig(('192.168.0.4', '255.255.255.0', '192.168.0.1', '8.8.8.8'))
+
+    .. method:: config('param')
+                config(param=value, ...)
+
+       Get or set general network interface parameters. These methods allow to work
+       with additional parameters beyond standard IP configuration (as dealt with by
+       ``ifconfig()``). These include network-specific and hardware-specific
+       parameters and status values. For setting parameters, keyword argument syntax
+       should be used, multiple parameters can be set at once. For querying, a
```

Comment:
> The documentation doesn't specify whether config() returns anything when called with keywords, so I would expect it to return None.

That's true. I'm sure CPython documentation doesn't list explicit return value for a lot of setter functions, as None is implied.

> I would expect a TypeError thrown on an invalid parameter, before any parameters were actually applied.

As in: name of the parameter? Perhaps raising TypeError would make sense, to follow behavior of "normal" keyword functions. However, currently esp8266 raises ValueError. However, "before any parameters were actually applied" is non-goal. All parameters are processed once in one loop, applied immediately and in row. That should be the expected behavior.

> We should explicitly call out whether parameters are set in the order passed.

I believe anything like that was impossible at all before CPython 3.6. Consider it underspecified for .config() call.

> If a user needs a specific order, they should probably make separate calls to config().

Yes.

> I would expect an exception if I tried to set one of the parameters to an invalid value, and that some, all or none of the other parameters may have been set before the exception.

Yes, with addition that invalid parameter name leads to the same effect, per above.